httpgrpc/server/server.go

```diff
@@ -37,16 +38,35 @@ func NewServer(handler http.Handler) *Server {
 	}
 }
 
+type nopCloser struct {
+	*bytes.Buffer
+}
+
+func (nopCloser) Close() error { return nil }
+
+func nopCloserBuffer(b *bytes.Buffer) io.ReadCloser {
+	return nopCloser{b}
+}
+
+// BytesBufferFromReader attempts to cast a `io.ReadCloser` as a `bytes.Buffer` to avoid reading and copying one that already exists.
+// In case of failure it will return nil and a false boolean.
+func BytesBufferFromReader(r io.ReadCloser) (*bytes.Buffer, bool) {
```

Comment:
I think the idiomatic Go way to do this would be as a method, which the caller can test via type assertion.

Reply:
In my opinion, I think this is simpler, what you described means I need to know about the type and the method. 

I'll do the change anyway, it'll save time for both of us.

Reply:
Just the method.  Example: https://github.com/grpc/grpc-go/blob/4be647f/rpc_util.go#L731-L733